src-ts/proposalStage.ts

```diff
@@ -118,23 +139,53 @@ export class GovernorQueueStage implements ProposalStage {
   public readonly identifier: string;
 
   public constructor(
-    public readonly proposalId: string,
-
-    public readonly target: string,
-    public readonly value: BigNumber,
-    public readonly callData: string,
+    public readonly targets: string[],
+    public readonly values: BigNumber[],
+    public readonly callDatas: string[],
     public readonly description: string,
 
     public readonly governorAddress: string,
-    public readonly signer: Signer
+    public readonly signerOrProvider: Signer | providers.Provider
   ) {
-    this.identifier = `${proposalId}:${target}:${value.toString()}:${callData}:${description}:${governorAddress}`;
+    this.identifier = keccak256(
+      defaultAbiCoder.encode(
+        ["address[]", "uint256[]", "bytes[]", "bytes32"],
+        [targets, values, callDatas, id(description)]
+      )
+    );
+  }
+
+  public static async extractStages(
+    receipt: TransactionReceipt,
+    arbOneSignerOrProvider: Provider | Signer
+  ): Promise<GovernorQueueStage[]> {
+    const govInterface = L2ArbitrumGovernor__factory.createInterface();
+    const proposalStages: GovernorQueueStage[] = [];
+    for (const log of receipt.logs) {
+      if (log.topics.find((t) => t === govInterface.getEventTopic("ProposalCreated"))) {
+        const propCreatedEvent = govInterface.parseLog(log)
+          .args as unknown as ProposalCreatedEventObject;
+
+        proposalStages.push(
+          new GovernorQueueStage(
+            propCreatedEvent.targets,
+            (propCreatedEvent as any)[3], // ethers is parsing an array with a single 0 big number as undefined, so we lookup by index
+            propCreatedEvent.calldatas,
+            propCreatedEvent.description,
+            log.address,
+            arbOneSignerOrProvider
+          )
+        );
+      }
+    }
+
+    return proposalStages;
   }
 
   public async status(): Promise<ProposalStageStatus> {
-    const gov = L2ArbitrumGovernor__factory.connect(this.governorAddress, this.signer);
+    const gov = L2ArbitrumGovernor__factory.connect(this.governorAddress, this.signerOrProvider);
 
-    const state = (await gov.state(this.proposalId)) as GovernorTimelockStatus;
+    const state = (await gov.state(this.identifier)) as GovernorTimelockStatus;
```

Comment:
Not introduced in this PR, but below - should the `GovernorTimelockStatus.Queued` case return `ProposalStageStatus.READY` (instead of `ProposalStageStatus.EXECUTED`?)